src/middlewared/middlewared/utils/io.py

```diff
@@ -1,24 +1,124 @@
+    UID = enum.auto()
+    GID = enum.auto()
+    PERMS = enum.auto()
+
+
+class UnexpectedFileChange(Exception):
+    def __init__(self, path, changes):
+        self.changes = changes
+        self.path = path
+        self.changes_str = ', '.join([x.name for x in FileChanges if self.changes & x])
+
+    def __str__(self):
+        return f'{self.path}: unexpected change in the following file attributes: {self.changes_str}'
+
+
+def write_if_changed(path, data, uid=0, gid=0, perms=0o755, dirfd=None, raise_error=False):
+    """
+    Commit changes to a configuration file.
+    `path` - path to configuration file. May be relative to a specified `dirfd`
+
+    `data` - expected file contents. May be bytes or string
+
+    `uid` - expected numeric UID for file
+
+    `gid` - expected numeric GID for file
+
+    `perms` - numeric permissions that file should have
+
+    `raise_error` - raise an exception if file ownership or permissions have unexpectedly
+    changed.
+    """
 
     if isinstance(data, str):
         data = data.encode()
 
-    if isinstance(path, int):
-        path = f'/proc/self/fd/{path}'
+    if not isinstance(perms, int):
+        raise ValueError('perms must be an integer')
+
+    if perms & ~(stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO):
+        raise ValueError(f'{perms}: invalid mode')
+
+    for xid in ((uid, 'uid'), (gid, 'gid')):
+        value, name = xid
+        if not isinstance(value, int):
+            raise ValueError(f'{name} must be an integer')
+
+        if value < 0 or value > ID_MAX:
+            raise ValueError(f'{name} must be between 0 and {ID_MAX}')
+
+    if dirfd is not None:
+        if not isinstance(dirfd, int):
+            raise ValueError('dirfd must be a valid file descriptor')
+
+        if not os.path.exists(f'/proc/self/fd/{dirfd}'):
+            raise ValueError(f'{dirfd}: file descriptor not found')
+
+    if dirfd is not None:
+        # tempfile API does not permit using a file descriptor
+        parent_dir = os.readlink(f'/proc/self/fd/{dirfd}')
+    else:
+        parent_dir = os.path.dirname(path)
+
+    changes = 0
+
+    try:
+        with open(os.open(path, os.O_RDONLY, dir_fd=dirfd), 'rb+') as f:
+            current = f.read()
+            if current != data:
+                changes |= FileChanges.CONTENTS
+
+            # The following cannot be skipped if we're changing file contents
+            # because we want accurate list of what has changed in file.
+
+            st = os.fstat(f.fileno())
+            if stat.S_IMODE(st.st_mode) != perms:
+                changes |= FileChanges.PERMS
+
+            if st.st_uid != uid:
+                changes |= FileChanges.UID
+
+            if st.st_gid != gid:
+                changes |= FileChanges.GID
+
+            if changes & (FileChanges.UID | FileChanges.GID):
+                os.fchown(f.fileno(), uid, gid)
+
+            if changes & FileChanges.PERMS:
+                os.fchmod(f.fileno(), perms)
```

Comment:
Could put these under an if to skip if contents have changed (to avoid doing the `fchown` or `fchmod` twice - on the file itself, and then on the temp file).

Reply:
I was thinking that if for some reason the tempfile write fails or rename fails we still want to correct the permissions and ownership on the original file.